app/src/main/kotlin/com/coredocker/android/views/components/user/UserCreateUpdateFragmentViewModel.kt

```diff
@@ -0,0 +1,118 @@
+package com.coredocker.android.views.components.user
+
+import android.os.Bundle
+import android.util.Log
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import com.coredocker.android.data.repository.DataRepository
+import com.coredocker.android.services.Navigate
+import com.coredocker.android.services.SnackBarNotification
+import com.coredocker.android.util.LoadingState
+import com.coredocker.android.util.extensions.default
+import kotlinx.coroutines.Dispatchers.IO
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.launch
+
+private val TAG: String = UserCreateUpdateFragmentViewModel::class.java.simpleName
+class UserCreateUpdateFragmentViewModel(
+    private val dataRepository: DataRepository,
+    private val navigate: Navigate,
+    private val notification: SnackBarNotification
+
+) : ViewModel() {
+    var id: String? = null
+    val name: MutableLiveData<String> = MutableLiveData()
+    val email: MutableLiveData<String> = MutableLiveData()
+    val password: MutableLiveData<String> = MutableLiveData()
+    val roles: MutableLiveData<List<String>> = MutableLiveData<List<String>>().default(
+        listOf("Guest")
+    )
+    val pageState: MutableLiveData<LoadingState> = MutableLiveData<LoadingState>().default(
+        LoadingState(false)
+    )
+
+    fun isChecked(role: String): Boolean {
+        return roles.value!!.contains(role)
+    }
+
+    val isUpdate: Boolean
+        get() {
+            return !this.id.isNullOrEmpty()
+        }
+
+    fun switchRoles(role: String) {
+        val mutateRoles = roles.value!!.toMutableList()
+        if (mutateRoles.contains(role)) {
+            mutateRoles.remove(role)
+        } else {
+            mutateRoles.add(role)
+        }
+        roles.postValue(mutateRoles)
+    }
+
+    fun onSaveClick() = GlobalScope.launch(IO) {
```

Comment:
👍 Hmm in this case I think I did this  for a reason, I did not want to tie it to the viewmodelScope because I was under the impression if the viewModel gets killed then it would send a cancel signal down the co-routine line. In this case I don't want that cancel routine to go all the way through. I still want my item to be saved even if the save is taking a long time.

I'm probably wrong around that scoping . Will run some tests to see if I get a clearer understanding of co-routine cancels.

Reply:
Your logic makes sense. Scoping and cancellation is definitely tricky in coroutines and it's part of their current push to formulate a simple, comprehensive and well documented process for that.

The problem is that if the viewModel gets ended but the GlobalScope is still running, the result will still attempt to be processed using viewModel functions, which means a memory leak. Or the GlobalScope is cleaned up forcibly by the garbagecollector, meaning it's tied to the viewModel lifecycle anyway. I'm not 100% sure what would happen here.

What I normally do in the case where I want something to run independent of the viewModel scope is to give the Repository a scope using `): CoroutineScope`, and then `launch` the first coroutine from there.

Then the ViewModel can trigger the refresh/apiCall in the repository and not wait for a result. Unfortunately to make that work you need to rely entirely on the repository's Flow<> data stream and have the viewModel be completely ignorant of database/api status. 

Some people recommend keeping a BroadcastChannel/Flow/LiveData<State> in your Repository that notifies all listeners that an error exists. This way the ViewModel can choose to notify the user if it is applicable. If you do this, you can have data in your database, but a failed api call while still having the ViewModel show the applicable ui due to listening to the dataStream and statusStream separately.

All of the things I said do add some boilerplate and trickiness, so I'm never sure if that truly is the best solution.

Let me know if you need any clarification.

Most of my assumptions are built upon the logic explained in this video, if you have spare time: https://www.youtube.com/watch?v=0IKHxjkgop4

It was made for RxJava instead of Kotlin, but it's considered on the of the great Android videos, and coroutines maps very well unto RxJava.